pages/partials/attachFilePanel.ejs

```diff
@@ -0,0 +1,99 @@
+          <a id="attachFileDeleteButton<%= iFile %>"
+             tabindex="0" class="btn btn-xs btn-secondary attachFileDeleteButton" role="button"
+             data-toggle="popover" data-container="body"
+             data-html="true" data-placement="auto" title="Confirm delete"
+             data-content="<%= include('attachFileDeleteForm',
+                           {id: 'attachFileDeleteButton' + iFile,
+                           variant: (typeof variant != 'undefined') ? variant : null,
+                           file}); %>"
+             data-trigger="manual" onclick="$(this).popover('show')">
+            <i class="far fa-trash-alt"></i>
+          </a>
+          <% } %>
+        </td>
+        <% } %>
+      </tr>
+      <% }); %>
+    <% } %>
+    </tbody>
+  </table>
+
+  <div class="card-footer">
+    <% if (!assessment_instance.open) { %>
+    <p class="small mb-0">
+      Attachments can't be added or deleted because the assessment is closed.
+    </p>
+    <% } else { /* assessment_instance is open */ %>
+    <!-- UPLOAD FILE -->
+    <button class="btn btn-xs btn-secondary" type="button" data-toggle="collapse" data-target="#attachFileCollapse" aria-expanded="false" aria-controls="attachFileCollapse">
+      Attach a file <i class="far fa-caret-square-down"></i>
+    </button>
+    <div class="collapse" id="attachFileCollapse">
+      <form class="attach-file-form" name="attach-file-form" method="POST" enctype="multipart/form-data" class="mb-3">
+        <p class="small mt-3">
+          Attached files will be saved here for your reference.
+        </p>
+        <div class="form-group">
+          <div class="custom-file">
+            <input type="file" name="file" class="custom-file-input" id="attachFileInput">
+            <label class="custom-file-label" for="attachFileInput">Choose file</label>
+          </div>
+        </div>
+        <div class="d-flex justify-content-end">
+          <div class="form-group mb-0">
+            <input type="hidden" name="__action" value="attach_file">
+            <% if (typeof variant != 'undefined') { %>
+            <input type="hidden" name="__variant_id" value="<%= variant.id %>">
+            <% } %>
+            <input type="hidden" name="__csrf_token" value="<%= __csrf_token %>">
+            <button type="submit" class="btn btn-primary">Attach file</button>
+          </div>
+        </div>
+      </form>
+    </div>
+
+    <!-- UPLOAD TEXT -->
+    <br>
+    <button class="btn btn-xs btn-secondary" type="button" data-toggle="collapse" data-target="#attachTextCollapse" aria-expanded="false" aria-controls="attachTextCollapse">
+      Attach text <i class="far fa-caret-square-down"></i>
+    </button>
+    <div class="collapse" id="attachTextCollapse">
+      <form method="POST" class="attach-text-form">
+        <p class="small mt-3">
+          Attached text will be saved here for your reference.
+        </p>
+        <input name="filename" value="notes.txt">
```

Comment:
Accessibility: filename and content inputs should have corresponding labels.

Reply:
I was worried that adding labels above these is just going to spread out the whole form vertically, and make it less clear what’s going on. I was hoping that the placeholder text in the content input was enough to clarify its purpose. As an example of using a placeholder rather than a label, see the “reply” box I’m typing in right now 😄

The filename input is a stronger case for a label, but I couldn’t really see where to put one and not be confusing. I’m open to suggestions on this!

Reply:
I'll compromise on an `aria-label` attribute on each (which the box I'm typing in now has 😛)

Reply:
Excellent point!